### PR TITLE
Create a passwordless ssh key in HA cluster

### PR DESCRIPTION
**Pull request by ldevulder**
Before bsc#1169581 crmsh used the default ssh root key (or created one) for the HA connection. Crmsh now uses a dedicated key for this.

As some tests need to do some passwordless ssh connections we now need to generate a pair. This was already done before in HA YaST test.

- Related ticket: N/A
- Needles: N/A
- Failing test: https://openqa.suse.de/tests/4281321#step/netweaver_network/11
- Verification run: TBD?

### CODE REVIEW

File: tests/ha/ha_cluster_join.pm [hunk 1]
```diff
@@ -35,6 +35,9 @@ sub run {
     diag 'Wait until cluster is initialized...';
     barrier_wait("CLUSTER_INITIALIZED_$cluster_name");
 
+    # Configure ssh key to enable ssh passwordless
+    get_root_ssh_key;
```
Comment (alvarocarvajald) [inline tests/ha/ha_cluster_join.pm:39]: Seems this is missing `$node_to_join`

File: tests/ha/yast_cluster_join.pm [hunk 2]
```diff
@@ -28,10 +28,7 @@ sub run {
     barrier_wait("CLUSTER_INITIALIZED_$cluster_name");
 
     # Configure ssh key to enable ssh passwordless
-    add_to_known_hosts($node_to_join);
-    assert_script_run "ls -altr /root/";
-    exec_and_insert_password("scp root\@$node_to_join:/root/.ssh/* /root/.ssh/");
-    assert_script_run "ssh root\@$node_to_join 'ssh-keyscan -H $hostname >> /root/.ssh/known_hosts'";
+    get_root_ssh_key;
```
Comment (alvarocarvajald) [inline tests/ha/yast_cluster_join.pm:31]: Seems this is missing `$node_to_join`

File: tests/ha/ha_cluster_join.pm (hunk 1 above)
Comment (ldevulder) [reply tests/ha/ha_cluster_join.pm:39]: Good catch!

File: tests/ha/ha_cluster_join.pm [hunk 3]
```diff
@@ -35,12 +35,16 @@ sub run {
...
+    # Configure ssh key to enable ssh passwordless
+    get_root_ssh_key $node_to_join;
+
     # Try to join the HA cluster through first node
     assert_script_run "ping -c1 $node_to_join";
     type_string "ha-cluster-join -yc $node_to_join ; echo ha-cluster-join-finished-\$? > /dev/$serialdev\n";
-    assert_screen 'ha-cluster-join-password', $join_timeout;
-    type_password;
-    send_key 'ret';
+    if (check_screen('ha-cluster-join-password', 5)) {
```
Comment (alvarocarvajald) [inline tests/ha/ha_cluster_join.pm:44]: In the previous version this waited for `$join_timeout` seconds for the password prompt but with the new version it'll wait for 5 seconds. Why the change?

File: tests/ha/remove_node.pm [hunk 4]
```diff
@@ -40,7 +40,14 @@ sub remove_state_join {
     is_node(2) ? script_run("$crm_mon_cmd", $default_timeout) : save_state;
 
     # Second node needs to be reintegrated
-    assert_script_run("$join_cmd $node_01", 2 * $join_timeout) if is_node(2);
+    if (is_node(2)) {
+        type_string "$join_cmd $node_01 ; echo ha-cluster-join-finished-\$? > /dev/$serialdev\n";
+        if (check_screen('ha-cluster-join-password', 5)) {
```
Comment (alvarocarvajald) [inline tests/ha/remove_node.pm:45]: Same question WRT the timeout value. Also, now that both this join command and the one in `tests/ha/ha_cluster_join` would expect for the password prompt, it's a good candidate to refactor the code as a function in `lib/hacluster`.

File: tests/ha/ha_cluster_join.pm (hunk 3 above)
Comment (ldevulder) [reply tests/ha/ha_cluster_join.pm:44]: Because we need the smallest value here (default is 0), so I would like to avoid scale out time here. The timeout here is just to check the password screen, not the timeout for the execution, this one is still in the `wait_serial` call.

File: tests/ha/remove_node.pm (hunk 4 above)
Comment (ldevulder) [reply tests/ha/remove_node.pm:45]: Yes, that's what I tried first but can take more time. Maybe after the BV?

File: tests/ha/ha_cluster_join.pm (hunk 3 above)
Comment (alvarocarvajald) [reply tests/ha/ha_cluster_join.pm:44]: That's the thing. Previous version waited at most `2 * $join_timeout`, while new one will wait at most `5 + $join_timeout`. More importantly, if we don't catch the password prompt with the `check_screen()` call, test will fail as bootstrap script will not finish waiting for the password. I understand the reasoning behind the 5 seconds as that should be more than enough for the password prompt to appear, and this will make test faster, but I worry about false positives in slower VMs. Guess we could always increase `TIMEOUT_SCALE` in settings in those cases though. I'll approve.

File: lib/hacluster.pm [hunk 5]
```diff
@@ -524,4 +526,20 @@ sub activate_ntp {
...
+    assert_script_run 'cp /root/.ssh/id_rsa.pub /root/.ssh/authorized_keys';
+}
+
+sub get_root_ssh_key {
+    my $node_to_join = shift;
+
+    # Configure ssh key to enable ssh passwordless
+    add_to_known_hosts($node_to_join);
+    exec_and_insert_password("scp root\@$node_to_join:/root/.ssh/* /root/.ssh/");
+    assert_script_run "ssh root\@$node_to_join 'ssh-keyscan -H ' . get_hostname . ' >> /root/.ssh/known_hosts'";
```
Comment (juadk) [inline lib/hacluster.pm:542]: The get_hostname doesn't work in the ssh call. I believe I had the same problem while I was doing the yast cluster test.
However, this solution is working in my [lab](http://1a102.qa.suse.de/tests/4361#step/ha_cluster_join/8) using the build with the bad crmsh patch:
```
sub get_root_ssh_key {
    my $node_to_join = shift;
    my $hostname     = get_hostname;
    # Configure ssh key to enable ssh passwordless
    add_to_known_hosts($node_to_join);
    exec_and_insert_password("scp root\@$node_to_join:/root/.ssh/* /root/.ssh/");
    assert_script_run "ssh -o StrictHostKeyChecking=no root\@$node_to_join 'ssh-keyscan -H $hostname >> /root/.ssh/known_hosts'";
}
```

Comment (alvarocarvajald) [reply lib/hacluster.pm:542]: Indeed. With `get_hostname` between double quotes, the method is not going to be called. Instead, the string `get_hostname` itself will be part of the shell command. Good catch!

Another option is to call it like this:
```
assert_script_run "ssh root\@$node_to_join 'ssh-keyscan -H '" . get_hostname . "' >> /root/.ssh/known_hosts'";
```
But I think it looks better with the `$hostname` variable.

File: tests/ha/ha_cluster_init.pm [hunk 6]
```diff
@@ -74,6 +74,9 @@ sub run {
     # Ensure that ntp service is activated/started
     activate_ntp;
 
+    # Generate ssh key
+    gen_root_ssh_key;
+
```
[the remark this answers is not in the document]
Comment (alvarocarvajald) [reply tests/ha/ha_cluster_init.pm:79]: Agree. But keep in mind `gen_ssh_root` is also copying the public key into the user's authorized_keys file. Perhaps you can add the check for `/root/.ssh/id_rsa` inside of `gen_root_ssh_key`.

[the remark this answers is not in the document]
Comment (juadk) [reply tests/ha/ha_cluster_init.pm:79]: yes agree but this is already done in `consoletest_setup.pm`
`assert_script_run('cat ~/.ssh/id_rsa.pub | tee -a ~/.ssh/authorized_keys');`

[the remark this answers is not in the document]
Comment (ldevulder) [reply tests/ha/ha_cluster_init.pm:79]: It is done in `consoletest_setup.pm` yes but was not always the case, so I think it's better to not check and erase the old key with a new one.
But anyway I can add a test to not doing this if needed. But the key of node02+ will be erase anyway with the one from node01.

File: lib/hacluster.pm (hunk 5 above)
Comment (ldevulder) [reply lib/hacluster.pm:542]: Thanks! I'm doing the change.

**Comment by ldevulder** [comment]
This PR is not needed anymore, so I closed it.
